### ISSUE TITLE
It seems grc versions are not updated on the server

### ISSUE THREAD

**Issue by Trondtr** [divvun/divvun-gramcheck-web#71]
For fao grc, the following sentence:

_Hann skal læra at virða tær skyldur og tey rættindi, sum fólkaræði byggir á, har ímillum teir sáttmálar, ið Sameindu Tjóðir hava samtykt í sambandi við barnarættindi og mannarættindi annars._ 

gets a correction "samtykt" to "samtykja", and the error message "Det ser ut til at verbet ditt ikkje skulle ende på ð.", corresponding to the tag &msyn-sup-inf. The problem is that this (false) alarm cannot be reproduced on the command line. The offending rule was removed from the grammarchecker.cg3 file on [february 6th](https://github.com/giellalt/lang-fao/commit/dc96a3e0e8f2fb582247e07c7f26c268a6a9447f).

The conclusion is thus that MS Word accesses an online grammar version dating from before February 6th. I do not know whether this holds for fao only, or whether it goes for all the online grammars, but it should be looked into. The fao grc is in use by a whole language community (the largest of the ones we serve) and the issue should thus get some priority. If the error carries over to the other grc versions the issue is even worse. Due to bug #70, I have only tested on MS Word.

**Comment by snomos** [comment]
This bug is really about how and when to update the grammar checker `*.zcheck` files on the production server. We have a standardised process for rolling out new spellers, but nothing for grammar checkers.

**Comment by Trondtr** [comment]
While waiting, we should document how to inspect the dates for the grammarcheckers.

**Comment by Trondtr** [comment]
After update on the server (today), I now check as describet above: I add the following sentence

*Hann skal læra at virða tær skyldur og tey rættindi, sum fólkaræði byggir á, har ímillum teir sáttmálar, ið Sameindu Tjóðir hava samtykt í sambandi við barnarættindi og mannarættindi annars.*

and send it to checkng. Now (contrary to yesterday) the setence comes out as **true negative.** The update thus is done, and the new version works.

The bug should not be closed, though: What happened today was that 3 languages were manually updated (fo.zcheck, nb, smn). The other languages were not updated, and the automatic update procedure is also not in place. Also, the documentation for how to check for dates (either via easter eggs or via inspecting the dates on pahkat.uit.no) is not in place.

The good news is that 3 languages are updated, so that the last 9 months of work is available for the language communities. Thanks!

**Comment by Trondtr** [comment]
> This bug is really about how and when to update the grammar checker `*.zcheck` files on the production server. We have a standardised process for rolling out new spellers, but nothing for grammar checkers.

So, let us partition the problem: if I understand correctly, we need two things:
1. a operational process for ensuring regression.
2. add the *.zcheck file to the list of files to be copied automatically to the server

There should be no reason for waiting months for the second point, it seems like a five-minute-fix, at most. The issue thus seems to have stranded at the first point. 

I do not whether speller update is linked to the `make check` procedure for transducers, but if it is, it should be straightforward to to the same for grammar checkers. We do have a test procedure, and it can be used for this purpose as well.

**Comment by unhammer** [comment]
I guess https://github.com/giellalt/lang-smj/issues/109#issuecomment-2616851380 is the same issue.

The version running on the server is from before September 2019 ( https://github.com/giellalt/lang-smj/commit/9e9a73a940de5a80c7149f73504fcc9aea6177a1 removed "msyn thingy" from errors.xml), that's over 5 years ago! @bbqsrc what's the procedure for getting these packages updated?

**Comment by snomos** [comment]
> The version running on the server is from before September 2019

That is very strange, and points to other issues. The `.zcheck` file for SMJ was sent to Gothenburg and uploaded to the server on December 17 (I think that included SMJ - two files were sent, but I am unable to tell what languages where updated then). And even if it was not at that time, I know for sure that the SMJ grammar checker has been updated the last time in 2024, not in 2019.

That is, there is something strange going on.

**Comment by snomos** [comment]
About the procedure: the present procedure goes as follows:

- compile a `.zcheck` file locally
- send the file to a person with access to the API server
- manually copy the file to the API server

@bbqsrc is working on an update to the whole Páhkat ecosystem that will include automatic releases of grammar checkers, the same way we do speller releases now (a version tag in a specific format). This update will also cover the api server, and thus make the api server update automatically upon stable/tagged releases.

**Comment by unhammer** [comment]
There is also a possibility that the errors.xml is of an older date than the rest of the checker (that the rest of the checker is from December, and it's just the errors.xml that's five years old). Even so, it would be good to have the latest changes in.

**Comment by snomos** [comment]
See my comments in https://github.com/giellalt/lang-smj/issues/109 - I am not able to replicate the issue with `errors.xml`.  I don't understand why you get what you get.

**Comment by unhammer** [comment]
errors.xml was https://github.com/divvun/divvun-gramcheck-web/issues/98 so not relevant for this.

> compile a .zcheck file locally

any reason not to just use the one built by Tino's packaging?

**Comment by snomos** [comment]
> any reason not to just use the one built by Tino's packaging?

No, not really, I just find it more convenient to build locally (=how do I get the one built by Tino?).

**Comment by unhammer** [comment]
well, any ubuntu machine doing `apt update && apt dist-upgrade` will have new ones in `/usr/share/voikko/4`, but I guess compiling is easier from a mac :) 

@TinoDidriksen is there a nightly-data script for osx? Feel like I've heard of one once.

**Comment by TinoDidriksen** [comment]
There is https://apertium.projectjj.com/osx/install-nightly-data.sh invoked as e.g. `./install-nightly-data.sh giella-smj-speller` which installs `/usr/local/share/voikko/4/smj.zcheck` and friends.

Rather simple script that downloads the .deb file from https://apertium.projectjj.com/osx/nightly/data.php and unpacks it, with some path adjusting for .mode files.

**Comment by snomos** [comment]
We are getting closer to what I would like to have:

- we have a beta API server (https://beta.api.giellalt.org/), which will soon get automatic updates of grammar checker files on each successful CI build after each push to GitHub
- for the production server, the target is still automated updates after tagged pushes, where the tag follows a certain format. Presently the update process is still manual.

Please note that the grammar checker servers (as well as the TTS servers) now uses [divvun-runtime](https://github.com/divvun/divvun-runtime) as the grammar checker/text processor, not libdivvun. It takes `*.drb` files, not `*.zcheck` files. Support for building `*.drb` files has been added lately, and will happen automatically if `divvun-runtime` is available to the build system, and grammar checker builds have been turned on.
